src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py

```diff
@@ -1097,11 +1094,18 @@ def _insert_opencl_init_routine(self, node):
         fortran_reader = FortranReader()
         container = fortran_reader.psyir_from_source(code)
         subroutine = container.children[0]
-        # Rename subroutine
-        subroutine.name = subroutine_name
+        sym_tab = subroutine.symbol_table.deep_copy()
+        generated_code = subroutine.pop_all_children()
+
+        real_subroutine = Routine(subroutine_symbol.name,
+                                  symbol=subroutine_symbol,
+                                  symbol_table=sym_tab,
+                                  symbol_tag="ocl_init_routine")
+        for nodes in generated_code:
+            real_subroutine.addchild(nodes)
```

Comment:
No, this is because the Routine doesn't expect its symbol to be attached to a symbol table when its created, so this is the workaround at the moment (I think the creation of this Routine object from Fortran source in the middle of the psyir was somewhat unexpected for me so I didn't have a good plan on how to handle it).

Reply:
> No, this is because the Routine doesn't expect its symbol to be attached to a symbol table when its created

But in this case we don't create it. The routine comes from the "psyir_from_source" above, so it is already attached and has a parent scope with a symbol table.

Reply:
Yes - but because the Routine can't expect the symbol to be attached to the symbol table the problem is the `tag` setting. The old code this wasn't an issue as it was adding the RoutineSymbol with `add_symbol` and providing a tag. This code is that same functionality but with the new system.

Reply:
There is an alternative option here (but I think its worse/handles more changes worse).

We can now instead do
```
subroutine.detach()                                                                                                                              node.addchild(subroutine)
sym = symtab.lookup(subroutine.name)
symtab.remove(sym)
symtab.add(sym, "ocl_init_routine")
```

The issue that could happen here is that the subroutine has no idea about whether its meant to have a tag, so if for some reason its detached and reattached it won't keep the tag. I'm not sure if this is a concern or not but I prefer the original implementation just in case.

Reply:
What if we reverted the routine creation to the original:
```
# Create the symbol for the routine and add it to the symbol table.
subroutine_symbol = symtab.new_symbol("psy_init",
                                       symbol_type=RoutineSymbol,
                                       tag="ocl_init_routine")
```
(note I removed the .name)

Add a routine.symbol property that only works when there is no _parent. As then we are free to change the symbol without breaking scopes.

and here just did:
```
subroutine.detach()
subroutine.symbol = subroutine_symbol
```

The tag will be kept, the replace_with tags will also be kept with my other suggestion? Would this allow us to remove the symbol_tag from the routine constructor?

Reply:
This would work here (though we still don't have a .symbol access, its still using _symbol for now unless you think its ok to allow this more widely). I'm just slightly wary of losing tag information if someone did 
```
routine.detach()
routine.attach(new_parent)
```

If the routine doesn't know anything about the tag and the user was expecting the tag to be carried over from the old parent to the new parent (for whatever reason) we'd lose it

Reply:
I am not sure I agree that this should keep the tag. The tag is a property of the scope, not the symbol. If we move a datasybol from one scope to another we also lose the tag. The method that intent on keeping the tag, e.g. copy or merge already take take to transfer it.

(If this behavior needs to be changed this is a separate issue and consistent for all symbols)